code/game/sound.dm

```diff
@@ -47,32 +47,38 @@ falloff_distance - Distance at which falloff begins. Sound is at peak volume (in
 
 	var/turf/turf_source = get_turf(source)
 
-	if (!turf_source)
+	if (!turf_source || !soundin || !vol)
 		return
 
 	var/maxdistance = (SOUND_RANGE + extrarange)
-	var/max_z_range = maxdistance / (MULTI_Z_DISTANCE + 1)
-
-	var/list/z_list = get_zs_in_range(turf_source.z, max_z_range)
+	var/source_z = turf_source.z
 
 	//allocate a channel if necessary now so its the same for everyone
 	channel = channel || SSsounds.random_available_channel()
 
  	// Looping through the player list has the added bonus of working for mobs inside containers
 	var/sound/S = sound(get_sfx(soundin))
-	var/list/listeners = list()
-	var/list/dead_listeners = list()
-	for(var/z in z_list)
-		listeners += SSmobs.clients_by_zlevel[z]
-		dead_listeners += SSmobs.dead_players_by_zlevel[z]
-	if(!ignore_walls) //these sounds don't carry through walls
-		listeners = listeners & hearers(maxdistance,turf_source)
-	for(var/mob/M as() in listeners)
-		if(get_dist(M, turf_source) <= maxdistance)
-			M.playsound_local(turf_source, soundin, vol, vary, frequency, falloff_exponent, channel, pressure_affected, S, maxdistance, falloff_distance, 1, use_reverb)
-	for(var/mob/M as() in dead_listeners)
-		if(get_dist(M, turf_source) <= maxdistance)
-			M.playsound_local(turf_source, soundin, vol, vary, frequency, falloff_exponent, channel, pressure_affected, S, maxdistance, falloff_distance, 1, use_reverb)
+	var/list/listeners = SSmobs.clients_by_zlevel[source_z].Copy()
+	/// Everyone that actually heard the sound
+	var/list/hearers = list()
+	var/turf/above_turf = SSmapping.get_turf_above(turf_source)
+	var/turf/below_turf = SSmapping.get_turf_below(turf_source)
+	if(ignore_walls)
+		if(above_turf && istransparentturf(above_turf))
+			listeners += SSmobs.clients_by_zlevel[above_turf.z]
+		if(below_turf && istransparentturf(turf_source))
+			listeners += SSmobs.clients_by_zlevel[below_turf.z]
+	else //these sounds don't carry through walls
+		listeners = get_hearers_in_view(maxdistance, turf_source)
+		if(above_turf && istransparentturf(above_turf))
+			listeners += get_hearers_in_view(maxdistance, above_turf)
+		if(below_turf && istransparentturf(turf_source))
+			listeners += get_hearers_in_view(maxdistance, below_turf)
+	for(var/mob/listening_mob in listeners | SSmobs.dead_players_by_zlevel[source_z])//observers always hear through walls
+		if(get_dist(listening_mob, turf_source) <= maxdistance)
+			listening_mob.playsound_local(turf_source, soundin, vol, vary, frequency, falloff_exponent, channel, pressure_affected, S, maxdistance, falloff_distance, 1, use_reverb)
+			hearers += listening_mob
+	return hearers
```

Comment:
get_hearers_in_view() is incredibly expensive, having it called multiple times is absolutely destroying the server via dview() calls

Reply:
so is this actually the issue or no